File: lib/models/project.dart
import 'enums.dart';

class Project {
  final String id;
  final String title;
  final String type;
  final int hours;
  final int budgetHours;
  final String clientName;
  final ProjectStatus status;
  final String? managerId;
  final String? teamId;

  Project({
    required this.id,
    required this.title,
    required this.type,
    required this.hours,
    required this.budgetHours,
    required this.clientName,
    required this.status,
    this.managerId,
    this.teamId,
  });

  factory Project.fromJoinedMap(Map<String, Object?> map) => Project(
        id: map['id'] as String,
        title: map['title'] as String,
        type: map['type'] as String,
        hours: map['hours'] as int,
        budgetHours: map['budget_hours'] as int,
        clientName: map['client_name'] as String,
        status: ProjectStatusDb.fromDb(map['status'] as String),
        managerId: map['manager_id'] as String?,
        teamId: map['team_id'] as String?,
      );
}

File: backend/DomainLayer/Domain/Projects/Project.cs
using DomainLayer.Domain.Builders;

namespace DomainLayer.Domain.Projects
{
    public class Project : ProjectBase
    {
        public double AllocatedHours { get; private set; }

        /// <summary>Hours logged / consumed against this project (budget is <see cref="AllocatedHours"/>).</summary>
        public double ConsumedHours { get; private set; }

        public ProjectStatus Status { get; private set; } = ProjectStatus.Active;

        public string ClientName { get; private set; } = string.Empty;

        public Guid ManagerId { get; private set; }

        public Guid TeamId { get; private set; }

        private Project()
        {
        }

        internal Project(
            Guid id,
            string title,
            DateTime startDate,
            DateTime endDate,
            int budgetHours,
            Guid managerId,
            Guid teamId,
            string clientName,
            ProjectStatus projectStatus,
            double consumedHours = 0)
            : base(id, title, startDate, endDate)
        {
            AllocatedHours = budgetHours;
            ConsumedHours = consumedHours;
            ManagerId = managerId;
            TeamId = teamId;
            ClientName = clientName;
            Status = projectStatus;
        }

        public void AddConsumedHours(double hours)
        {
            if (hours <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(hours), "Hours must be positive.");
            }

            ConsumedHours += hours;
        }

        public void ChangeStatus(ProjectStatus newStatus)
        {
            Status = newStatus;
        }

        public override double GetTotalAllocatedHours()
        {
            // Implementação simples para calcular os dias
            return AllocatedHours;
        }

        /// <summary>Starts fluent construction; call <see cref="ProjectBuilder.Build"/> when complete.</summary>
        public static ProjectBuilder Builder() => new ProjectBuilder();
    }
}
